arcconfig/arcconfig.go

```diff
@@ -0,0 +1,164 @@
+// the config loading and validation functions.
+package arcconfig
+
+import "time"
+
+// APIRuntimeControlConfig is an Apigee Environment-level config for
+// Envoy Adapter. It contains a list of operations for the adapter to
+// perform request authentication and authorization.
+type APIRuntimeControlConfig struct {
+	// Name of the config
+	Name string `yaml:"name"`
+
+	// Revision of the config
+	Revision string `yaml:"revision"`
+
+	// A list of Operations, names of which must be unique within the config.
+	Operations []*Operation `yaml:"operations,omitempty"`
+}
+
+// An API Operation associates a set of rules with a set of request matching
+// settings.
+type Operation struct {
+	// Name of the operation. Unique within a APIRuntimeControlConfig.
+	Name string `yaml:"name"`
+
+	// Authentication defines the AuthenticationRequirement
+	Authentication AuthenticationRequirement `yaml:"authentication,omitempty"`
+
+	// Authorization defines the ConsumerAuthorization
+	Authorization ConsumerAuthorization `yaml:"authorization,omitempty"`
+
+	// HTTP matching rules for this operation.
+	// The path templates within which should be fully denormalized.
+	HTTPMatches []HTTPMatch `yaml:"http_match,omitempty"`
+
+	// Name of the target server for this operation. This will be sent to Envoy
+	// for routing to the corresponding upstream cluster upon a successful
+	// authorization of the operation.
+	Target string `yaml:"target"`
+}
+
+// AuthenticationRequirement defines the authentication requirement.
+// Precisely one of JWT, Any and All should be set.
+type AuthenticationRequirement struct {
+	// JWT defines the JWTAuthentication
+	JWT JWTAuthentication `yaml:"jwt,omitempty"`
+
+	// Any contains a list of AuthenticationRequirements.
+	// A successful authentication requires one of them being satisfied.
+	Any []AuthenticationRequirement `yaml:"any,omitempty"`
+
+	// All contains a list of AuthenticationRequirements.
+	// A successful authentication requires all of them being satisfied.
+	All []AuthenticationRequirement `yaml:"all,omitempty"`
+}
+
+// JWTAuthentication defines the JWT authentication
+type JWTAuthentication struct {
+	// Name of this JWT Provider, unique within the Proxy.
+	Name string `yaml:"name"`
+
+	// JWT issuer ("iss" claim)
+	Issuer string `yaml:"issuer"`
+
+	// A remote JWKS source
+	RemoteJWKS RemoteJWKS `yaml:"remote_jwks"`
+
+	// Audiences contains a list of audiences
+	Audiences []string `yaml:"audiences,omitempty"`
+
+	// Header name that will contain decoded JWT payload in requests forwarded to
+	// target.
+	ForwardPayloadHeader string `yaml:"forward_payload_header,omitempty"`
+
+	// Locations where JWT may be found. First match wins.
+	In []HTTPParameter `yaml:"in"`
+}
+
+// RemoteJWKS contains information for remote JWKS
+type RemoteJWKS struct {
+	// URL of the JWKS
+	URL string `yaml:"url"`
+
+	// CacheDuration of the JWKS
+	CacheDuration time.Duration `yaml:"cache_duration,omitempty"`
```

Comment:
What does the YAML form of this look like?

Reply:
For example, `30s` or `1h`.

Reply:
FYI, this is the standard for Go. https://pkg.go.dev/maze.io/x/duration#ParseDuration